src/shred_network/shred_receiver.zig

```diff
@@ -186,13 +201,338 @@ pub const ShredReceiver = struct {
             },
         );
         self.metrics.passed_to_inserter_count.add(self.shred_batch.len);
+
+        try self.handleDuplicateSlots(allocator, &result);
+
         result.deinit();
 
         self.metrics.batch_size.observe(self.shred_batch.len);
     }
 
     const MAX_SHREDS_PER_ITER = 1024;
 
+    /// Handles detected duplicate slots by:
+    /// - Send duplicate slot notifications to be handled in consensus part of replay
+    /// - Store the duplicate proof in the ledger
+    /// - Broadcast the duplicate proof via gossip
+    /// Analogous to [check_duplicate](https://github.com/anza-xyz/agave/blob/aa2f078836434965e1a5a03af7f95c6640fe6e1e/core/src/window_service.rs#L155)
+    fn handleDuplicateSlots(
+        self: *ShredReceiver,
+        allocator: Allocator,
+        result: *const ShredInserter.Result,
+    ) !void {
+        const sender = self.params.duplicate_slots_sender orelse return;
+        if (result.duplicate_shreds.items.len == 0) return;
+
+        for (result.duplicate_shreds.items) |duplicate_shred| {
+            switch (duplicate_shred) {
+                .Exists => |shred| {
+                    const shred_slot = shred.commonHeader().slot;
+                    // Unlike the other cases we have to wait until here to decide to handle the duplicate and store
+                    // in ledger. This is because the duplicate could have been part of the same insert batch,
+                    // so we wait until the batch has been written.
+                    if (try self.params.ledger_reader.isDuplicateSlot(shred_slot)) {
+                        continue; // A duplicate is already recorded, skip
+                    }
+
+                    const existing_shred_payload =
+                        (try self.params.ledger_reader.isShredDuplicate(allocator, shred)) orelse
```

Comment:
You don't actually need these parens, the precedence ought to work out to the equivalent of `try foo orelse continue` = `foo catch |e| return e orelse continue`.